File: py/tests/test_unicontext.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 lituus-lab
from unicontext import BUDGET_MAX, BUDGET_MIN, abi_version, valid_budget, version


def test_version_matches_the_manifest():
    assert version() == "1.0.0"
    assert abi_version() == 1


def test_budget_bounds_come_from_the_header():
    assert (BUDGET_MIN, BUDGET_MAX) == (128, 32768)


def test_budget_contract_matches_the_bounds():
    assert valid_budget(BUDGET_MIN)
    assert valid_budget(BUDGET_MAX)
    assert not valid_budget(BUDGET_MIN - 1)
    assert not valid_budget(BUDGET_MAX + 1)
    assert not valid_budget(0)


def test_out_of_c_range_is_answered_not_raised():
    assert not valid_budget(2**63)
    assert not valid_budget(-(2**63))

File: tests/c/test_unicontext.c
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 lituus-lab
#include "UniContext.h"

#include <stdio.h>
#include <string.h>

static int failures = 0;

static void check(int condition, const char *message) {
  if (condition) {
    printf("ok   %s\n", message);
  } else {
    printf("FAIL %s\n", message);
    failures++;
  }
}

int main(void) {
  check(unicontext_abi_version() == 1, "the ABI generation is 1");

  const char *version = unicontext_version();
  check(version != NULL, "the version is returned");
  check(version != NULL && strcmp(version, UNICONTEXT_VERSION) == 0,
        "the library agrees with the header's version string");

  /* The bounds are inclusive, and one step outside each is refused. */
  check(unicontext_valid_budget(UNICONTEXT_BUDGET_MIN) == 1,
        "the smallest budget is accepted");
  check(unicontext_valid_budget(UNICONTEXT_BUDGET_MAX) == 1,
        "the largest budget is accepted");
  check(unicontext_valid_budget(UNICONTEXT_BUDGET_MIN - 1) == 0,
        "one below the smallest is refused");
  check(unicontext_valid_budget(UNICONTEXT_BUDGET_MAX + 1) == 0,
        "one above the largest is refused");

  /* The ABI never raises: extreme input is answered, not rejected. */
  check(unicontext_valid_budget(0) == 0, "zero is refused");
  check(unicontext_valid_budget(-2147483647 - 1) == 0, "INT_MIN is refused");
  check(unicontext_valid_budget(2147483647) == 0, "INT_MAX is refused");

  if (failures > 0) {
    printf("%d check(s) failed\n", failures);
    return 1;
  }
  printf("all checks passed\n");
  return 0;
}
